torch/csrc/distributed/c10d/init.cpp

```diff
@@ -148,6 +162,27 @@ class PythonStore : public ::c10d::Store {
     return std::vector<uint8_t>(str.begin(), str.end());
   }
 
+  // Note: this function manually calls the Python-side overload
+  // for this function instead of using the PYBIND11_OVERLOAD_XYZ
+  // macros. This is done so that the Python-side function can
+  // return a py::bytes instead of a std::vector<uint8_t>.
+  std::vector<uint8_t> compareSet(
+      const std::string& key,
+      const std::vector<uint8_t>& currentValue,
+      const std::vector<uint8_t>& newValue) override {
+    pybind11::gil_scoped_acquire gil;
+    pybind11::function fn = pybind11::get_overload(
+        static_cast<const ::c10d::Store*>(this), "compare_set");
+    TORCH_INTERNAL_ASSERT(fn);
+    // Cast return value from Python to py::bytes, then implicitly
+    // convert that to a std::string, so that we can construct a
+    // std::vector<uint8_t>. There is no API for directly accessing
+    // the contents of the py::bytes object.
+    std::string str =
+        pybind11::cast<py::bytes>(fn(key, currentValue, newValue));
```

Comment:
What is the exact return type of `fn`? These two lines seem highly inefficient to me. We copy the return value of `fn` to a temporary `std::string` just for the sake of constructing a `std::vector<>`. If we can have a way to access the data pointer and the size from the return value of `fn`, we can avoid constructing the temporary string object.

Reply:
For context, PythonStore was originally created to be used as an extension of the store class, so the return value would be whatever the user who is defining the extension implements.

However, it is quite old and I don't believe it is actually being used anymore. Still, I needed to add a compareSet for PythonStore in order to make the Store::compareSet a pure virtual function. I will follow up and find out if it is being used anymore and can potentially remove the entire class in a later PR. I will leave this implementation as is for now.